core/src/main/java/org/apache/iceberg/encryption/EncryptionUtil.java

```diff
@@ -70,31 +76,105 @@ public static KeyManagementClient createKmsClient(Map<String, String> catalogPro
     return new BaseEncryptedOutputFile(encryptingOutputFile, EncryptionKeyMetadata.empty());
   }
+
+  public static EncryptionKeyMetadata setFileLength(
+      EncryptionKeyMetadata encryptionKeyMetadata, long manifestListLength) {
+    Preconditions.checkState(
+        encryptionKeyMetadata instanceof StandardKeyMetadata,
+        "Cant set file length in %s",
+        encryptionKeyMetadata.getClass());
+    ((StandardKeyMetadata) encryptionKeyMetadata).setFileLength(manifestListLength);
+    return encryptionKeyMetadata;
+  }
+
+  private static long parseFileLength(ByteBuffer keyMetadataBuffer) {
+    StandardKeyMetadata standardKeyMetadata = StandardKeyMetadata.parse(keyMetadataBuffer);
+    return standardKeyMetadata.fileLength();
+  }
+
+  public static void getKekCacheFromMetadata(FileIO io, Map<String, KeyEncryptionKey> kekCache) {
+    Preconditions.checkState(
+        io instanceof EncryptingFileIO,
+        "Can't set KEK cache - IO %s is not instance of EncryptingFileIO",
+        io.getClass());
+    EncryptionManager encryption = ((EncryptingFileIO) io).encryptionManager();
+    Preconditions.checkState(
+        encryption instanceof StandardEncryptionManager,
+        "Can't set KEK cache - encryption manager %s is not instance of StandardEncryptionManager",
+        encryption.getClass());
+    ((StandardEncryptionManager) encryption).addKekCache(kekCache);
+  }
+
+  public static InputFile decryptManifestListFile(
+      ManifestListFile manifestListFile, FileIO fileIO) {
+    Preconditions.checkArgument(
+        fileIO instanceof EncryptingFileIO,
+        "Cannot read manifest list (%s) because it is encrypted but the configured "
+            + "FileIO (%s) does not implement EncryptingFileIO",
+        manifestListFile.location(),
+        fileIO.getClass());
+    EncryptingFileIO encryptingFileIO = (EncryptingFileIO) fileIO;
+
+    Preconditions.checkArgument(
+        encryptingFileIO.encryptionManager() instanceof StandardEncryptionManager,
+        "Cannot read manifest list (%s) because it is encrypted but the "
+            + "encryption manager (%s) is not StandardEncryptionManager",
+        manifestListFile.location(),
+        encryptingFileIO.encryptionManager().getClass());
+    StandardEncryptionManager standardEncryptionManager =
+        (StandardEncryptionManager) encryptingFileIO.encryptionManager();
+
+    KeyEncryptionKey keyEncryptionKey =
+        standardEncryptionManager.keyEncryptionKey(manifestListFile.metadataEncryptionKeyID());
+
+    Ciphers.AesGcmDecryptor keyDecryptor = new Ciphers.AesGcmDecryptor(keyEncryptionKey.key());
+    ByteBuffer manifestListKeyMetadata =
+        ByteBuffer.wrap(
+            keyDecryptor.decrypt(
+                ByteBuffers.toByteArray(manifestListFile.encryptedKeyMetadata()), null));
+    ((BaseManifestListFile) manifestListFile).setDecryptedKeyMetadata(manifestListKeyMetadata);
```

Comment:
What is the purpose of keeping the encryption key in the `ManifestListFile` after it has been decrypted?

I don't think this should be casting the instance.

Reply:
SGTM, we can drop this.